### PR TITLE
ascanrules: FP fix for postgres timebased fp

### PR DESCRIPTION
**Pull request by itsdivyanshjain**
## Overview
Hey zap team, this is another time based FP fix for postgresql injection. Here solution is similar to previously raised PRs(https://github.com/zaproxy/zap-extensions/pull/4679). Also, this branch is checkout from that nosql branch. So, before merging this expecting that above mentioned PR to be merged.

Regards,
Divyansh Jain (from astra security)

## Checklist
- [ ] Update help
- [X] Update changelog
- [X] Run `./gradlew spotlessApply` for code formatting
- [X] Write tests
- [X] Check code coverage
- [X] Sign-off commits
- [X] Squash commits
- [X] Use a descriptive title

For more details, please refer to the [developer rules and guidelines](https://www.zaproxy.org/docs/developer/dev-rules-and-guidelines/).

### CODE REVIEW

**Comment by itsdivyanshjain** [comment]
> Please remove the unrelated commits.

This utilized updated timing utils which is from nosql fp fix branch that's why you are seeing unrelated commits. It would be better to wait for nosql fp fix MR to get merged & then compare this. Also, if you still want to check only this two file is changed `addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java`  `addOns/ascanrules/src/test/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRuleUnitTest.java`

**Comment by kingthorin** [comment]
I see that now, it'll just have to be rebased after the other is done I guess.

**Comment by thc202** [comment]
Changed to WIP pending merge of #4679.

**Comment by thc202** [comment]
Removed the WIP, the timing utils were already migrated, needs rebase to pick the latest changes.

**Comment by itsdivyanshjain** [comment]
@kingthorin @thc202 review this too

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 1]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
+                        // use TimingUtils to detect a response to sleep payloads
+                        isInjectable =
+                                TimingUtils.checkTimingDependence(
+                                        BLIND_REQUEST_LIMIT,
+                                        sleepInSeconds,
+                                        requestSender,
+                                        TIME_CORRELATION_ERROR_RANGE,
+                                        TIME_SLOPE_ERROR_RANGE);
+                    } catch (SocketException ex) {
+                        LOGGER.info(
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:310]: `debug`

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 2]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
+                        // We Found IT!
+                        String extraInfo =
+                                Constant.messages.getString(
+                                        "astraScripts.ascanrules.sqlinjection.alert.timebased.extrainfo",
+                                        payloadValue,
+                                        "",
+                                        paramValue);
+                        String attack =
+                                Constant.messages.getString(
+                                        "astraScripts.ascanrules.sqlinjection.alert.booleanbased.attack",
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:328]: Keys need update.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 3]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
+                                        paramName,
+                                        payloadValue);
+                        LOGGER.info(
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:331]: `debug`

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 4]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
+                                "[Time Based Postrgres SQL Injection -  Found] on parameter [{}] with value [{}]",
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:332]: Remove extra space after the hyphen.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 5]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
+                                .setConfidence(Alert.CONFIDENCE_MEDIUM)
+                                .setName(getName() + " - Time Based")
+                                .setUri(getBaseMsg().getRequestHeader().getURI().toString())
+                                .setParam(paramName)
+                                .setAttack(attack)
+                                .setMessage(message.get())
+                                .setOtherInfo(extraInfo)
+                                .raise();
+                    }
+                } catch (SocketTimeoutException ex) {
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:346]: Remove this one, already handled above.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 6]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
-
-                // send it.
+                                .replace(SLEEP_TOKEN, String.valueOf(sleepInSeconds));
+                TimingUtils.RequestSender requestSender =
+                        x -> {
+                            HttpMessage timedMsg = getNewMsg();
+                            message.set(timedMsg);
+                            String finalPayload =
+                                    SQL_POSTGRES_TIME_REPLACEMENTS[index]
+                                            .replace(ORIG_VALUE_TOKEN, paramValue)
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:291]: `payloadValue`

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 7]
```diff
@@ -259,129 +269,91 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
             for (int timeBasedSQLindex = 0;
                     timeBasedSQLindex < SQL_POSTGRES_TIME_REPLACEMENTS.length
                             && doTimeBased
                             && countTimeBasedRequests < doTimeMaxRequests;
                     timeBasedSQLindex++) {
-                HttpMessage msgAttack = getNewMsg();
-                String newTimeBasedInjectionValue =
-                        SQL_POSTGRES_TIME_REPLACEMENTS[timeBasedSQLindex]
+                countTimeBasedRequests++;
+                int index = timeBasedSQLindex;
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:279]: Should not be needed.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java (hunk 2 above)
Comment (thc202) [reply addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java]: Needs further tweaks.

File: addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java [hunk 8]
```diff
@@ -259,129 +268,85 @@ public void scan(HttpMessage originalMessage, String paramName, String paramValu
...
-                            .setOtherInfo(extraInfo)
-                            .setMessage(msgAttack)
-                            .raise();
-
-                    LOGGER.debug(
-                            "A likely Time Based SQL Injection Vulnerability has been found with [{}] URL [{}] on field: [{}]",
-                            msgAttack.getRequestHeader().getMethod(),
-                            msgAttack.getRequestHeader().getURI(),
-                            paramName);
-                    return;
```
Comment (thc202) [inline addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/SqlInjectionPostgreScanRule.java:381]: This one should be restored, to not keep trying with other payloads.

**Review by psiinon — APPROVED** [review]

**Pull request merged**